### PR TITLE
Combine ip and ip2 libraries

### PR DESCRIPTION
**Pull request by kgerheiser**
I think this PR is finally ready.

This PR combines the ip and ip2 libraries into a single library with modern interfaces and modules. Internally, instead of passing around the complete grib1/grib2 spec, an abstract `ip_grid_descriptor` object is used as a wrapper. Then, all the interpolation routines can re-use the majority of the code. Spectral interpolation is a special case and under the hood there are still two separate routines for grib1/grib2 because there's a lot of case specific things going on that were difficult to generalize.

The code has been tested with a series of unit tests (Intel and GCC, with 66%+ coverage), UFS_UTILS, and wgrib2.

To use the code `use ip_mod` which provides access to the main routines `ipolates`,  `ipolatev`, and `gdswzd`. The specific routine can also be called by appending `_grib1` or `_grib2`.

To Do: Finish the Doxygen documentation. A lot of it is complete, but still some more to go.

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
@kgerheiser Can you build it on Hera or Dell? I can test it with UFS_UTILS.

**Comment by kgerheiser** [comment]
Try this `export ip_ROOT=/home/Kyle.Gerheiser/NCEPLIBS-ip/build/install` on Hera

**Review by edwardhartnett — APPROVED** [review]
Taking this on faith, which is not good. Next time, break up changes into smaller PRs...

**Comment by kgerheiser** [comment]
@GeorgeGayno-NOAA I ran the `snow2mdl` and `global_cycle` tests on Hera, and it passed. Do any of the others exercise this code? 

UFS_UTILS also passes all the Github workflows with the updated library.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA I ran the `snow2mdl` and `global_cycle` tests on Hera, and it passed. Do any of the others exercise this code?
> 
> UFS_UTILS also passes all the Github workflows with the updated library.

What UFS_UTILS branch are you using?

**Comment by kgerheiser** [comment]
https://github.com/kgerheiser/UFS_UTILS/tree/feature/combined-ip, which is up-to-date with develop

**Comment by edwardhartnett** [comment]
@kgerheiser you could make a GitHub workflow for UFS_UTILS that uses the develop branch of all the dependent libraries. Then, not only will we check the current state of play, but we will find out if we break anything...

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA I ran the `snow2mdl` and `global_cycle` tests on Hera, and it passed. Do any of the others exercise this code?
> 
> UFS_UTILS also passes all the Github workflows with the updated library.

The 'orog' code uses IPLIB.

**Comment by kgerheiser** [comment]
@GeorgeGayno-NOAA is that included in one of the consistency tests?

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA is that included in one of the tests?

The only tests we have for the 'orog' code currently are the 'grid_gen' consistency tests.

**Comment by kgerheiser** [comment]
`grid_gen` tests pass as well.

**Comment by GeorgeGayno-NOAA** [comment]
The corresponding UFS_UTILS issue is: https://github.com/NOAA-EMC/UFS_UTILS/issues/242

**Comment by GeorgeGayno-NOAA** [comment]
@kgerheiser I checked out your UFS_UTILS feature/combined-ip branch on Hera. It compiled, but was pointing to the current stack version of IPLIB. Am I looking at the wrong branch? I recall having to make source code changes to get the combined iplib to work with `snow2mdl`. See https://github.com/NOAA-EMC/UFS_UTILS/issues/242

**Comment by kgerheiser** [comment]
Oh, I didn't commit that change. I removed `module load ip` and put `setenv ip_ROOT /home/Kyle.Gerheiser/NCEPLIBS-ip/build/install`

File: docs/user_guide.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+# Documentation of the general interpolation library - iplib
 
 ## Introduction
 
-The NCEP general interpolation library 2 (ip2lib) contains Fortran 90 subprograms
+The NCEP general interpolation library (ip2lib) contains Fortran 90 subprograms
 to be used for interpolating between nearly all grids used at NCEP.
 The library is particularly efficient when interpolating many fields at one time.
-The library has been extensively tested with AIX and Intel Fortran compilers.
+The library has been extensively tested with GNU and Intel Fortran compilers.
```
Comment (edwardhartnett) [inline docs/user_guide.md]: Well I would strike this line. I'm not sure the library has been "extensively tested" anywhere...

File: src/ip_gaussian_grid_mod.f90 [hunk 2]
```diff
@@ -1,3 +1,6 @@
+!> @file
+!! @brief Gaussian grid
+
```
Comment (edwardhartnett) [inline src/ip_gaussian_grid_mod.f90:3]: Author tag at file level as well...

File: src/ip_gaussian_grid_mod.f90 [hunk 3]
```diff
@@ -25,6 +28,12 @@ module ip_gaussian_grid_mod
 
 contains
 
+  !> Initializes a gaussian grid given a grib1 grib1_descriptor object.
+  !! 
+  !! @param[inout] self The grid to initialize
+  !! @param[in] g1_desc A grib1_descriptor
+  !!
+  !! @author Kyle Gerheiser
```
Comment (edwardhartnett) [inline src/ip_gaussian_grid_mod.f90:36]: Add a @date tag as well.

File: src/ip_gaussian_grid_mod.f90 [hunk 4]
```diff
@@ -110,128 +124,128 @@ subroutine init_grib2(self, g2_desc)
 
   end subroutine init_grib2
 
+  !$$$  SUBPROGRAM DOCUMENTATION BLOCK
```
Comment (edwardhartnett) [inline src/ip_gaussian_grid_mod.f90:127]: This whole comment block must also be doxygenated.

File: src/ip_interpolator_mod.f90 [hunk 5]
```diff
@@ -1,3 +1,5 @@
+!> @file
+!! @brief
```
Comment (edwardhartnett) [inline src/ip_interpolator_mod.f90]: That's too brief! Put in a one-sentence description of the file. Also author and date tags (if known).

File: src/ipolates.f90 [hunk 6]
```diff
@@ -21,6 +29,7 @@ module ipolates_mod
 
 contains
 
+
```
Comment (edwardhartnett) [inline src/ipolates.f90]: Need a doxygen documentation block for every subroutine.

File: src/ipolates.f90 [hunk 7]
```diff
@@ -95,293 +104,293 @@ SUBROUTINE IPOLATES_grib1(IP,IPOPT,KGDSI,KGDSO,MI,MO,KM,IBI,LI,GI, &
     call ipolates_grid(ip, ipopt, grid_in, grid_out, mi, mo, km, ibi, li, gi, no, rlat, rlon, ibo, lo, go, iret)
 
   END SUBROUTINE IPOLATES_GRIB1
-  
 
+
+  !$$$  SUBPROGRAM DOCUMENTATION BLOCK
+  !
```
Comment (edwardhartnett) [inline src/ipolates.f90]: This and all other such documentation blocks must be converted to doxygen. Lower-case it as well.

**Review by edwardhartnett — CHANGES_REQUESTED** [review]

**Review by edwardhartnett — APPROVED** [review]
If you've got this working, probably you should merge and continue with doxygen work in a new PR...

File: src/ip_interpolators_mod.f90 [hunk 8]
```diff
@@ -1,19 +1,33 @@
...
+module ip_interpolators_mod
+  use bilinear_interp_mod
+  use bicubic_interp_mod
+  use budget_interp_mod
+  use neighbor_interp_mod
+  use spectral_interp_mod
+  use neighbor_budget_interp_mod
   implicit none
 
+  !> @param Constant to choose BILINEAR interpolation method
```
Comment (edwardhartnett) [inline src/ip_interpolators_mod.f90:16]: The param statements all go into the doxygen header at the top of the subroutine.

File: src/ipolates.f90 [hunk 9]
```diff
@@ -46,17 +68,23 @@ subroutine ipolates_grid(ip, ipopt, grid_in, grid_out, mi, mo, km, ibi, li, gi,
 
     select case(ip)
     case(BILINEAR_INTERP_ID)
-       CALL interpolate_bilinear(IPOPT,grid_in,grid_out,MI,MO,KM,IBI,LI,GI,NO,RLAT,RLON,IBO,LO,GO,IRET)
+       CALL interpolate_bilinear(IPOPT,grid_in,grid_out,MI,MO,KM,IBI&
+            &,LI,GI,NO,RLAT,RLON,IBO,LO,GO,IRET)
     case(BICUBIC_INTERP_ID)
-       CALL interpolate_bicubic(IPOPT,grid_in,grid_out,MI,MO,KM,IBI,LI,GI,NO,RLAT,RLON,IBO,LO,GO,IRET)
+       CALL interpolate_bicubic(IPOPT,grid_in,grid_out,MI,MO,KM,IBI&
```
Comment (edwardhartnett) [inline src/ipolates.f90:74]: For easy review, what I suggest is to do all whitespace only reformatting as a separate PR, with no code changes. So merge this PR, then reformat the code however you like, and then do a PR with that.

That way we can easily scan the whitespace changes to ensure no code was accidently changed.

**Review by edwardhartnett — APPROVED** [review]

File: src/ip_interpolators_mod.f90 (hunk 8 above)
Comment (kgerheiser) [reply src/ip_interpolators_mod.f90:16]: It's not a subroutine. `@param` can be used to document module variables and constants.

Comment (edwardhartnett) [reply src/ip_interpolators_mod.f90:16]: OK, interesting. Can we get this merged and then discuss doxygen afterwards?
